rfcs/0000-ckb-softfork-activation/0000-ckb-softfork-activation.md

```diff
@@ -0,0 +1,132 @@
+
+### States
+
+With each epoch and softfork, we associate a deployment state. The possible states are:
+
+1. DEFINED is the first state that each softfork starts. The genesis epoch is by definition in this state for each deployment.
+2. STARTED for epochs past the `start_epoch`.
+3. LOCKED_IN for one epoch after the first epoch period with STARTED epochs of which at least `threshold` has the associated bit set in `version`.
+4. ACTIVE for all epochs after the LOCKED_IN epoch.
+5. FAILED for one epoch period past the `timeout_epoch`, if LOCKED_IN was not reached.
+
+### Bit flags
+
+The `version` block header field is to be interpreted as a 32-bit little-endian integer (as present), and bits are selected within this integer as values (1 << N) where N is the bit number.
+
+Blocks in the STARTED state get a `version` whose bit position bit is set to 1. The top 3 bits of such blocks must be 001, so the range of possible `version` values is [0x20000000...0x3FFFFFFF], inclusive.
+
+By restricting the top 3 bits to 001, we get 29 out of those for this proposal and support future upgrades for different mechanisms. When a block `version` does not have top bits 001, it is treated as if all bits are 0 for deployments.
+
+Miners should continue setting the bit in the LOCKED_IN phase, so uptake is visible, though this does not affect consensus rules.
+
+### New consensus rules
+The new consensus rules for each softfork are enforced for each block with an ACTIVE state.
+
+### State transitions
+
+![State transitions](images/state-transitions.png)
+
+The genesis epoch has a state DEFINED for each deployment, by definition.
+
+```rust
+if epoch.is_genesis() {
+    return ThresholdState::DEFINED;
+}
+```
+
+We remain in the initial state until we reach the `start_epoch`.
+
+```rust
+match state {
+    ThresholdState::DEFINED => {
+        if epoch.number() >= start {
+            next_state = ThresholdState::STARTED;
+        }
+    }
+```
+
+After an epoch in the STARTED state, we tally the bits set and transition to LOCKED_IN if a sufficient number of blocks in the past epoch set the deployment bit in their version numbers. If the threshold has not been met and we reach the `timeout_epoch`, we transition directly to FAILED.
+
+Note that a block's state never depends on its version, only on that of its ancestors.
+
+```rust
+match state {
+    ThresholdState::STARTED => {
+        let mut count = 0;
+        for block in (0..epoch_length) {
+            if (block.version() & 0xE0000000 == 0x20000000 && (block.version() >> bit) & 1 == 1) {
+                ++count;
+            }
+        }
+        let threshold_number = threshold_number(epoch_length, threshold);
+        if count >= threshold_number {
+            next_state = ThresholdState::LOCKED_IN;
+        } else if epoch_ext.number() >= timeout {
+            next_state = ThresholdState::FAILED;
+        }
+    }
+```
```

Comment:
Is this check for `LOCKED_IN` run at the beginning of each epoch, or at each block?

Reply:
The state of the blocks in the same epoch is the same, and the state of the epoch depends on the previous epoch, i.e. we only count the previous epoch. So this check will only be done once per epoch and we cache cache the results.